Added support for theme folder

olivercox opened the pull request:
If options.settings['theme'] is set by calling

"app.set('theme', __dirname + '/themes/' + storeConfig.ActiveTheme);"

then all templates will be loaded from the themes folder if they exist.
If they don't exist templates will be loaded as usual.

Commit:
Added support for theme folder in renderFile

If options.settings['theme'] is set by calling

"app.set('theme', __dirname + '/themes/' + storeConfig.ActiveTheme);"

then all templates will be loaded from the themes folder if they exist.
If they don't exists templates will be loaded as usual.

RandomEtc commented on index.js:
```diff
@@ -82,6 +82,14 @@ var renderFile = module.exports = function(file, options, fn){
+    //Get the file path relative to the theme folder by replacing the options.settings.views path with the options.settings['theme'] path
+    var themeFilePath = dirname(file).replace(resolve(options.settings.views), resolve(options.settings['theme']))
+    //If the current file exists reletive to the theme folder set the file to that path
+    if (fs.existsSync(resolve(themeFilePath, basename(file)))) file = resolve(themeFilePath, basename(file));
```
I think the call to `existsSync` should be cached in production apps.

olivercox replied:
Yes that makes sense. Not sure how to do this but will take a look. 
Could the current view cache be used?

olivercox replied:
Would something like this work for the current cache

<pre><code>
var engine = options.settings['view engine'] || 'ejs'
  , desiredExt = '.' + engine
  , ext = extname(file) || desiredExt
  , key = [themeFilePath, file, ext].join('-');
//Check if the file has already been resolved in the theme folder to avoid io
if (options.cache && cache[key]) file = cache[key]; 
if (fs.existsSync(resolve(themeFilePath, basename(file)))) {
    cache[key] = resolve(themeFilePath, basename(file));
    file = resolve(themeFilePath, basename(file));
}</code><pre>

RandomEtc commented:
This looks simple enough - can you add some tests maybe an example? Apologies up-front that the tests are a little bit of a rats' nest, feel free to start a separate file.

Commit:
Added caching to theme folder support

Added the use of the cache object to store templates that have already
been resolved in the theme folder

bebraw commented:
+1

Any progress with this?

By the way as a workaround it would be awesome if relative paths worked. Seems to fail with good old `..`.
